packages/governance/src/contractGovernance/governFilter.js

```diff
@@ -0,0 +1,117 @@
+// @ts-check
+
+import { E } from '@endo/eventual-send';
+import { Far } from '@endo/marshal';
+import { fit, keyEQ } from '@agoric/store';
+import { M } from '@agoric/vat-data';
+
+import {
+  ChoiceMethod,
+  QuorumRule,
+  ElectionType,
+  coerceQuestionSpec,
+} from '../question.js';
+
+const { details: X, quote: q } = assert;
+
+/**
+ * Make a pair of positions for a question about whether to update the offer
+ * filter. If the vote passes, the list of blocked invitation strings will be
+ * updated.
+ *
+ * @param {string[]} strings
+ */
+const makeOfferFilterPositions = strings => {
+  const positive = harden({ strings });
+  const negative = harden({ dontUpdate: strings });
+  return { positive, negative };
+};
+
+/**
+ * Setup to allow governance to block some invitations.
+ *
+ * @param {ERef<ZoeService>} zoe
+ * @param {Instance} governedInstance
+ * @param {ERef<TimerService>} timer
+ * @param {() => Promise<PoserFacet>} getUpdatedPoserFacet
+ * @param {GovernorFacet} governorFacet
+ * @returns {Promise<FilterGovernor>}
+ */
+const setupFilterGovernance = async (
+  zoe,
+  governedInstance,
+  timer,
+  getUpdatedPoserFacet,
+  governorFacet,
+) => {
+  /** @type {WeakSet<Instance>} */
+  const voteCounters = new WeakSet();
+
+  /** @type {VoteOnOfferFilter} */
+  const voteOnFilter = async (voteCounterInstallation, deadline, strings) => {
+    fit(strings, M.arrayOf(M.string()));
+    const { positive, negative } = makeOfferFilterPositions(strings);
+
+    /** @type {OfferFilterIssue} */
+    const issue = harden({ strings });
+    const questionSpec = coerceQuestionSpec({
+      method: ChoiceMethod.UNRANKED,
+      issue,
+      positions: [positive, negative],
+      electionType: ElectionType.OFFER_FILTER,
+      maxChoices: 1,
+      closingRule: { timer, deadline },
+      quorumRule: QuorumRule.MAJORITY,
+      tieOutcome: negative,
+    });
+
+    const { publicFacet: counterPublicFacet, instance: voteCounter } = await E(
+      getUpdatedPoserFacet(),
+    ).addQuestion(voteCounterInstallation, questionSpec);
+
+    voteCounters.add(voteCounter);
+
+    // CRUCIAL: Here we wait for the voteCounter to declare an outcome, and then
+    // attempt to invoke the API if that's what the vote called for. We need to
+    // make sure that outcomeOfUpdateP is updated whatever happens.
+    //
+    // * If the vote passed, invoke the API, and return the positive position
+    // * If the vote was negative, return the negative position
+    // * If we can't do either, (the vote failed or the API invocation failed)
+    //   return a broken promise.
```

Comment:
Can the caller distinguish between the two kinds of failures? If that's supported part of the API, let's document it in JSDoc. If it isn't, let's say so explicitly.

Reply:
A promise is either resolved (with a value) or rejected (with a [reason](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch#syntax)). So it should be possible to provide different reasons for these two kinds of failure.

What I don't know is whether we want to commit to that as part of this API. Leaving that to y'all but my request is that the decision be documented.

Reply:
First a bit of terminology: There are two ways to use an error to indicate something went wrong: throw, or using the error as a rejected-promise-reason. Usually, including here, we don't care to distinguish these so I'll use the term "raise error" to mean both.

To grok this question in its fullness, we must first distinguish two possible purposes of the information carried by a raised error.
   * Something went wrong. To provide an out-of-band human programmer with diagnostic information to help them diagnose a possible bug in some code somewhere. 
   * To enable code to test-and-branch on something, such that subsequent computation proceeds differently depending on this information carried by the error. We need to distinguish two subcases:
      * Use to express non-local non-erroneous control flow
      * To react differently to different kinds or serverities of what went wrong. Depending on what went wrong, we may wish to take more severe action programmatically, like terminating a vat or shutting down the chain.

One unfortunate feature of JavaScript creates an additional complication, which we cope with by using another unfortunate feature of JavaScript.
   * JavaScript has no non-local non-erroneous exit operation suitable for use as non-erroneous control-flow.
   * JavaScript allow non-errors to be raised (thrown or used as promise rejections).

For non-local non-erroneous control flow, our recommended practice (which we've only recently decided on and started practicing) is that normal non-erroneous control flow, not indicating a problem that should be brought to human attention, should happen by raising a non-error. Then, the code that catches something and branches is not asking "Is this an error I expect or something else?" but rather "Is this a non-error I expect or something else?". In both cases, if it is something else the normal reaction will be to rethrow as if the intermediary were not in the way. 

Let's take the prescriptive stance that, for production code, only the ""Is this a non-error I expect or something else?" conditional is legitimate. Non-diagnostic production code should never have a conditional of the form "Is this an error I expect or something else?"

This leaves the final case, where different kinds of things going wrong should result in different severities of recovery action. For these, generally, the answer is to report the error in some way other than raising it, such as a call to `zcf.shutdownWithFailure`. An important cross-over case is what happens if an error is thrown during the execution of an atomic block after the commit point. In this case, the catcher escalates the severity of a thrown error into a shutdown. But the catcher does this escalation unconditionally --- again, without inspecting the information conveyed by the error.

So now, returning to our distinct error conditions
> (the vote failed or the API invocation failed)

I covered with @Chris-Hibbert what each means, and both indicate something went wrong, such that diagnostic info would likely be useful to help a programmer figure it out. Neither represents a non-erroneous control flow condition. Neither calls for a higher severity reaction, like sudden vat termination. Thus, non-diagnostic code should not distinguish between them, and the API spec should not imply that there is any such distinction for code to test.

This leaves only the question about whether this particular piece of code, responsible for propagating both of these errors, should bother to intercept one or both in order to enhance the diagnostic information it carries, for example, by using `assert.note`. This is always an esthetic judgement call. But my sense from talking to @Chris-Hibbert is there is nothing about the distinction between these two cases that isn't adequately conveyed by their stack traces. 

> Leaving that to y'all but my request is that the decision be documented.

@Chris-Hibbert can I leave it to you to see that this explanation be moved to a suitable place? Thanks.

Reply:
> can I leave it to you to see that this explanation be moved to a suitable place? 

Yes. There's room for it in the Wiki.